### PR TITLE
parking lot class

### PR DESCRIPTION
**Pull request by dem4ply**
add the class with the logic to handle the parking lot

### CODE REVIEW

**Comment by carlosheiras17** [comment]
Could you please adjust your code according to Python’s PEP 8 coding standard?

File: parking_lot/factories.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+
+tariffs = [ 'hourly', 'daily' ]
+
+
+class Car( factory.Factory ):
+    car = factory.Faker( 'license_plate' )
+    tariff = fuzzy.FuzzyChoice( tariffs )
+    location = factory.Faker( 'pyint' )
+    hourly_fee = factory.Faker( 'pyint' )
```
Comment (carlosheiras17) [inline parking_lot/factories.py]: what is going to happen when we have 100 new tariffs?

Class Car:
  x_tariff = bla
  z_tariff = ble
  ... 
  .
  .
  n_tariff

File: parking_lot/factories.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+class Car_add_hourly( factory.Factory ):
+    car = factory.Faker( 'license_plate' )
+    tariff = 'hourly'
+
+    class Meta:
+        model = dict
+
+
+class Car_add_daily( factory.Factory ):
```
Comment (carlosheiras17) [inline parking_lot/factories.py]: if you create one class per tariff then this is going to explode in size when adding 100 tariffs leading to difficult to maintain files.

File: parking_lot/parking_lot.py [hunk 3]
```diff
@@ -1,8 +1,150 @@
...
+
+    def exit( self ):
+        self.finish = datetime.datetime.now()
+
+    @property
+    def fee( self ):
+        if self.is_free_charge:
+            return 0
+
+        if self.is_daily:
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: this is not going to scale well with multiple kind of tariffs

if a:
if b:
if c:
..
.
.
if z:

File: parking_lot/parking_lot.py [hunk 4]
```diff
@@ -1,8 +1,150 @@
...
+        self.hourly_fee = hourly_fee
+        self.daily_fee = daily_fee
+
+    @property
+    def amount_lot( self ):
+        return self._amount_lot
+
+    @amount_lot.setter
+    def amount_lot( self, value ):
+        self._amount_lot = value
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py:58]: why self._amount_lot is not declared during __init__ ? isn't supposed all attributes of an instance are declared on init method?

File: parking_lot/parking_lot.py [hunk 5]
```diff
@@ -1,8 +1,150 @@
...
+
+
+class Parking_lot:
+    locations = None
+    lot = None
+
+    def __init__( self, amount, hourly_fee=1, daily_fee=20 ):
+        self.lot = {}
+        self.locations = []
+        self.amount_lot = amount
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py:48]: this is already a @property, why to declare it twice? just to trigger the setter method? if yes, then is like telling your object to build itself, instead of building the object externally with a factory method, or something similar.

File: parking_lot/parking_lot.py [hunk 6]
```diff
@@ -1,8 +1,150 @@
...
+    def proportional_diff_time( self ):
+        diff_time = self.finish - self.start
+        if self.is_hourly:
+            return ( diff_time.total_seconds() // 3600 ) + 1
+        if self.is_daily:
+            return ( diff_time.total_seconds() // 86400 ) + 1
+
+
+class Parking_lot:
+    locations = None
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: what is the purpose of this and below variable as class attributes?

File: parking_lot/parking_lot.py [hunk 7]
```diff
@@ -1,8 +1,150 @@
...
+        return len( self.lot ) >= self.amount_lot
+
+    @property
+    def is_empty( self ):
+        return (
+            sum( ( 1 for i in self.locations if i is None ) )
+            == self.amount_lot )
+
+    def find_next_available_location( self ):
+        return self.locations.index( None )
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: this raise a value error when parking lot is full

File: parking_lot/parking_lot.py [hunk 8]
```diff
@@ -1,8 +1,150 @@
...
+        self.locations[ location ] = result
+        self.lot[ car ] = result
+        return result
+
+    def remove( self, location ):
+        car = self.locations[ location ]
+        if car is None:
+            raise Cannot_find_car_error( location )
+        self.locations[ location ] = None
+        del self.lot[ car.car ]
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: car.car is confusing, why not car.license_plate ?

File: tests/test_parking_lot.py [hunk 9]
```diff
@@ -1,16 +1,206 @@
...
+            car.fee, car.proportional_diff_time * car.hourly_fee )
+
+    def test_daily_tariff_should_be_proportional_after_15min( self ):
+        car = Car_daily_factory.build()
+        self.assertEqual(
+            car.fee, car.proportional_diff_time * car.hourly_fee )
+
 
+class Test_Parking_lot( unittest.TestCase ):
+    amount = 10
```
Comment (carlosheiras17) [inline tests/test_parking_lot.py:42]: why to declare amount as class variable?

File: parking_lot/parking_lot.py [hunk 10]
```diff
@@ -1,8 +1,150 @@
...
+        car.exit()
+        return car
+
+    @property
+    def is_full( self ):
+        return (
+            sum( ( 1 for i in self.locations if i is not None ) )
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: why to sum all the "not None" positions instead of consulting the len(self.lot) ?

File: parking_lot/parking_lot.py [hunk 11]
```diff
@@ -1,8 +1,150 @@
...
+            == self.amount_lot )
+        return len( self.lot ) >= self.amount_lot
+
+    @property
+    def is_empty( self ):
+        return (
+            sum( ( 1 for i in self.locations if i is None ) )
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: could you please simplify this to avoid overhead, this is doable in o(n) time, even o(1) if you use extra variables to count how many ocuppied lots you have.

File: parking_lot/parking_lot.py [hunk 12]
```diff
@@ -1,8 +1,150 @@
...
+
+        location = self.find_next_available_location()
+        result = Car(
+            car=car, tariff=tariff, location=location,
+            hourly_fee=self.hourly_fee, daily_fee=self.daily_fee )
+        self.locations[ location ] = result
+        self.lot[ car ] = result
+        return result
+
+    def remove( self, location ):
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: why to remove by a location instead of license_plate?

File: parking_lot/parking_lot.py [hunk 13]
```diff
@@ -1,8 +1,150 @@
...
+        car = self.locations[ location ]
+        if car is None:
+            raise Cannot_find_car_error( location )
+        self.locations[ location ] = None
+        del self.lot[ car.car ]
+        car.exit()
+        return car
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py:83]: what is the purpose behind returning car on this remove method? initially you must have the car to trigger this function because is the only place where location is injected.

File: README.rst [hunk 14]
```diff
@@ -4,3 +4,34 @@ parking_lot
...
+
+	from parking_lot.parking_lot import Parking_lot
+	# create parking lot with 10 spaces
+	parking_lot = Parking_lot( amount=10 )
+	car = parking_lot.add( car='X774HY98', tariff='hourly' )
+	car = parking_lot.remove( car.location )
+	# print parking fee
+	print( car.fee )
+	car = parking_lot.add( car='X774HY98', tariff='daily' )
+	car = parking_lot.remove( car.location )
```
Comment (carlosheiras17) [inline README.rst]: parking_lot.remove receives the location inside a car object and return the same car again.

File: README.rst [hunk 15]
```diff
@@ -4,3 +4,34 @@ parking_lot
...
+	# print parking fee
+	print( car.fee )
```
Comment (carlosheiras17) [inline README.rst]: why car have a fee, instead of a ticket entity having a fee, or third entity generating fees based on a ticket's metadata?

File: parking_lot/parking_lot.py [hunk 16]
```diff
@@ -1,8 +1,150 @@
...
+    @property
+    def is_empty( self ):
+        return (
+            sum( ( 1 for i in self.locations if i is None ) )
+            == self.amount_lot )
+
+    def find_next_available_location( self ):
+        return self.locations.index( None )
+
+    def __contains__( self, index ):
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: could you please show me where are these magic methods being used? (__getitem__ and __iter__ too)

File: parking_lot/parking_lot.py [hunk 17]
```diff
@@ -1,8 +1,150 @@
...
+            'status': 'error',
+            'error': message
+        }
+
+
+class Car:
+    def __init__(
+            self, car, tariff, location, hourly_fee, daily_fee,
+            start=None, finish=None ):
+        self.car = car
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: a car should have a license_plate instead of a car.

File: parking_lot/parking_lot.py [hunk 18]
```diff
@@ -1,8 +1,150 @@
...
+
+    def __init__( self, location=None ):
+        message = self.message.format( location )
+        self.data = {
+            'status': 'error',
+            'error': message
+        }
+
+
+class Car:
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: this class feels more like a ticket because it have entry and exit timestamps, license_plate of the car, tariff, the metadata needed for calculating fees on car's exit.

File: parking_lot/parking_lot.py [hunk 19]
```diff
@@ -1,8 +1,150 @@
...
+    @property
+    def fee( self ):
+        if self.is_free_charge:
+            return 0
+
+        if self.is_daily:
+            return self.proportional_diff_time * self.daily_fee
+        if self.is_hourly:
+            return self.proportional_diff_time * self.hourly_fee
+        raise NotImplementedError(
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: this exception have another purpose according documentation: https://docs.python.org/3/library/exceptions.html#NotImplementedError

File: parking_lot/parking_lot.py [hunk 20]
```diff
@@ -1,8 +1,150 @@
...
+
+    @property
+    def is_hourly( self ):
+        return self.tariff == 'hourly'
+
+    @property
+    def proportional_diff_time( self ):
+        diff_time = self.finish - self.start
+        if self.is_hourly:
+            return ( diff_time.total_seconds() // 3600 ) + 1
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: could you please explain me why is needed to sum 1 always?

File: parking_lot/parking_lot.py [hunk 21]
```diff
@@ -1,8 +1,150 @@
...
+        if start is None:
+            start = datetime.datetime.now()
+        self.start = start
+        self.finish = finish
+
+        self.hourly_fee = hourly_fee
+        self.daily_fee = daily_fee
+
+    def exit( self ):
+        self.finish = datetime.datetime.now()
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py:26]: I think Car class is doing to much, it calculates fee, but also knows location of where car is stored, knows about the different kind of tariffs and car exit timestamp too.

File: parking_lot/parking_lot.py [hunk 22]
```diff
@@ -1,8 +1,150 @@
...
-from flask import Flask
+import datetime
 
-app = Flask(__name__)
 
-@app.route("/")
-def hello_world():
-    return "<p>Hello, World!</p>"
+class Parking_lot_error( Exception ):
+    message = 'unhandle error with parking lot'
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: the purpose of having a class attribute is to share it with others classes and instances, but in this case every class is overriding it, do we really need this?

File: parking_lot/parking_lot.py [hunk 23]
```diff
@@ -1,8 +1,150 @@
...
+
+
+class Full_parking_error( Parking_lot_error ):
+    message = 'No free space.'
+
+
+class Cannot_find_car_error( Parking_lot_error ):
+    message = 'Cannot find a car in the location {}.'
+
+    def __init__( self, location=None ):
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: this init method is similar to parent class one, could you please refactor this in order to use super().__init__()

File: tests/test_parking_lot.py [hunk 24]
```diff
@@ -1,16 +1,206 @@
 #!/usr/bin/env python
 # -*- coding: utf-8 -*-
+import datetime
+import random
+import unittest
+from parking_lot.parking_lot import (
+    Parking_lot, Full_parking_error, Car, Cannot_find_car_error
+)
+from parking_lot.factories import (
```
Comment (carlosheiras17) [inline tests/test_parking_lot.py]: is module factories only used for test purposes? if yes, why not to relocate the file to tests folder to don't distribute code intended for test code to client side.

File: parking_lot/parking_lot.py [hunk 25]
```diff
@@ -1,8 +1,150 @@
...
+
+    @property
+    def amount_lot( self ):
+        return self._amount_lot
+
+    @amount_lot.setter
+    def amount_lot( self, value ):
+        self._amount_lot = value
+        self.locations = [ None ] * self._amount_lot
+        for v in self.lot.values():
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py:60]: what is this code intended for ? is the purpose to adjust size of the parking lot on runtime?

File: parking_lot/parking_lot.py [hunk 26]
```diff
@@ -1,8 +1,150 @@
...
+            self.locations[ v.location ] = v
+
+    def add( self, car, tariff ):
+        if self.is_full:
+            raise Full_parking_error
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: could you please explain me why to raise an error when parking lot is full?

**Review by carlosheiras17 — CHANGES_REQUESTED** [review]

File: parking_lot/parking_lot.py [hunk 27]
```diff
@@ -1,8 +1,150 @@
...
+        if self.is_daily:
+            return self.proportional_diff_time * self.daily_fee
+        if self.is_hourly:
+            return self.proportional_diff_time * self.hourly_fee
+        raise NotImplementedError(
+            f"the tariff {self.tariff} is not implemented" )
+
+    @property
+    def is_free_charge( self ):
+        diff_time = self.finish - self.start
```
Comment (carlosheiras17) [inline parking_lot/parking_lot.py]: you are calculating at least twice this same value, why not turn it into a property as well?

**Review by carlosheiras17 — CHANGES_REQUESTED** [review]

File: parking_lot/factories.py (hunk 1 above)
Comment (dem4ply) [reply parking_lot/factories.py]: for that case is going to use different classes or function is going to depend of the new tariff

File: parking_lot/factories.py (hunk 2 above)
Comment (dem4ply) [reply parking_lot/factories.py]: those class are only to be used with unit testing, is not the intention to be used for the logic of the parking lot.

File: parking_lot/parking_lot.py (hunk 3 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: i know, but is enough for two tariff, to build a better way to calculate the fee i going to need more tariffs or some vision on how going to need implemented.

File: parking_lot/parking_lot.py (hunk 4 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py:58]: originally was created in the init, but i change that to make easy some unit tests.

File: parking_lot/parking_lot.py (hunk 5 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py:48]: is not declare it twice is triggering the setter, this class is not using factories or some external way to create, is creating itself.

File: parking_lot/parking_lot.py (hunk 6 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: i put those in the phase of prototyping is not been used.

File: parking_lot/parking_lot.py (hunk 7 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: you are right, i add the unit test and raise a more expected exception

File: parking_lot/parking_lot.py (hunk 8 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: i chose car.car because in the example of the homework the input and the output key say car

File: tests/test_parking_lot.py (hunk 9 above)
Comment (dem4ply) [reply tests/test_parking_lot.py:42]: the amount was intended to be use inheritance in the unit tests.

File: parking_lot/parking_lot.py (hunk 10 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: you are right, thanks

File: parking_lot/parking_lot.py (hunk 11 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: i dont think is necessary to use extra variables, but i simplified with your previous commentary with property is full.

File: parking_lot/parking_lot.py (hunk 12 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: the logic with the rest api, remove the car using the location and not the license plate.

File: parking_lot/parking_lot.py (hunk 13 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py:83]: the returning the car is because i going to need the car data in the rest api to return response and in the api is only going to have the location and not the license plate or the car object

File: README.rst (hunk 14 above)
Comment (dem4ply) [reply README.rst]: yes, is the intention in the logic, i going to change the example to another to make more sense.

File: README.rst (hunk 15 above)
Comment (dem4ply) [reply README.rst]: when i start the prototype i was not thinking in tickets but car in the parking lot, perhaps ticket is a better metaphor.

File: parking_lot/parking_lot.py (hunk 16 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: is only used in the tests 

**containts**
https://github.com/dem4ply/parking_lot/pull/1/files#diff-448fa941f2114beb179c4ba0fc8454018d46d998802ccade748bdbad994711e6R93

**getitem**
https://github.com/dem4ply/parking_lot/pull/1/files#diff-448fa941f2114beb179c4ba0fc8454018d46d998802ccade748bdbad994711e6R109

the **iter** is needed in the rest api
https://github.com/dem4ply/parking_lot/blob/feature/rest_api/parking_lot/parking_lot_api.py#L40

File: parking_lot/parking_lot.py (hunk 17 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: i use **car** because was the key in the rest api, i going to do the change because make more sense.

File: parking_lot/parking_lot.py (hunk 19 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: partially true, here is been used to indicate where is need to implement new  tariff.

File: parking_lot/parking_lot.py (hunk 18 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: yes, you are right, ticket can be a better metaphor, i was thinking in car inside of the parking lot.

File: parking_lot/parking_lot.py (hunk 20 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: the plus 1 add the fee of the first partial hour or day after the first 15 minutes.

File: parking_lot/parking_lot.py (hunk 21 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py:26]: i put the logic of the fee and handle the time inside of the class Car in case of a edge case like the fee change in runtime and need to respect the fee of the previous cars.

File: parking_lot/parking_lot.py (hunk 22 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: the message is for the rest api and is changing in each because change the message of the exception but not other parts.

File: tests/test_parking_lot.py (hunk 24 above)
Comment (dem4ply) [reply tests/test_parking_lot.py]: i normally leave the factories in the package because normally i need those in another projects, but i can do the change.

File: parking_lot/parking_lot.py (hunk 25 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py:60]: yes, is in the case of the edge case if need to change in the rest api and was helpful in the tests

File: parking_lot/parking_lot.py (hunk 26 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: the rest api catch the exception and return the expected message and the method **add** dont have a good way to communicate with the user/developer that is full.

File: parking_lot/parking_lot.py (hunk 27 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: right i dont notice that, thanks.

**Comment by denis-kishenko** [comment]
MAJOR: Seems I can add the same car multiple times
`
$ curl --noproxy '*'  'http://127.0.0.1:8080/add?car=X774HY98&tariff=hourly'
{"car": "X774HY98", "tariff": "hourly", "status": "success", "location": 1, "start": "2024-05-02 11:00:45"}

curl --noproxy '*'  'http://127.0.0.1:8080/add?car=X774HY98&tariff=hourly'
{"car": "X774HY98", "tariff": "hourly", "status": "success", "location": 2, "start": "2024-05-02 11:00:46"}
`

**Comment by denis-kishenko** [comment]
MEDIUM: Could you please fix class names according to PEP8 https://peps.python.org/pep-0008/#class-names

File: parking_lot/parking_lot.py (hunk 6 above)
Comment (denis-kishenko) [reply parking_lot/parking_lot.py]: MEDIUM: Hi @dem4ply, could you please remove all unused variables. It could be confusing for later maintenance.

**Comment by denis-kishenko** [comment]
MAJOR: Could you please fix these errors, service successfully processes those requests.
http://127.0.0.1:8080/add?car=&tariff=hourly
http://127.0.0.1:8080/add?car=X774HY98&tariff=hourlyXXX
Wrong parameters will cause unnecessary tickets from customers for our team

File: parking_lot/parking_lot.py (hunk 19 above)
Comment (denis-kishenko) [reply parking_lot/parking_lot.py]: MAJOR: As I understand, NotImplementedError is uncontrolled error and will be processed by error handler as service error while it is user's fault. If it is correct let's fix this. We will be spammed with user complains about service errors while really it user's fault.

File: parking_lot/factories.py (hunk 1 above)
Comment (denis-kishenko) [reply parking_lot/factories.py]: MAJOR: Could you please implement pluggable tariffs to minimize base code changes as much as possible.

File: parking_lot/parking_lot.py (hunk 6 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: no problem

File: parking_lot/parking_lot.py (hunk 19 above)
Comment (dem4ply) [reply parking_lot/parking_lot.py]: done

**Comment by dem4ply** [comment]
fix the next issues

MEDIUM: Hi @dem4ply, could you please remove all unused variables. It could be confusing for later maintenance.

MAJOR: As I understand, NotImplementedError is uncontrolled error and will be processed by error handler as service error while it is user's fault. If it is correct let's fix this. We will be spammed with user complains about service errors while really it user's fault.

MAJOR: Seems I can add the same car multiple times

MEDIUM: Could you please fix class names according to PEP8

MAJOR: Could you please fix these errors, service successfully processes those requests.
http://127.0.0.1:8080/add?car=&tariff=hourly
http://127.0.0.1:8080/add?car=X774HY98&tariff=hourlyXXX

**Comment by denis-kishenko** [comment]
Should we expect a new code or a task to be completed?

File: parking_lot/factories.py (hunk 1 above)
Comment (dem4ply) [reply parking_lot/factories.py]: done

**Comment by dem4ply** [comment]
> Should we expect a new code or a task to be completed?

yes, i got busy before finish the last issue but it's done

MAJOR: Could you please implement pluggable tariffs to minimize base code changes as much as possible.
